chain/network_block.go

```diff
@@ -0,0 +1,536 @@
+func (s *NetworkBlockSource) Start() {
+	log.Debugf("Starting network block provider")
+	s.wg.Add(1)
+	go s.requestHandler()
+}
+
+// Stop stops the NetworkBlockSource from accepting any more incoming block
+// requests.
+func (s *NetworkBlockSource) Stop() {
+	close(s.quit)
+	s.wg.Wait()
+
+	// Close all client channel to signal we can no longer satisfy their
+	// request.
+	for _, clients := range s.activeReqs {
+		for _, client := range clients {
+			// Since a single client can be registered for multiple
+			// blocks, we'll need to ensure we don't close their
+			// channel twice. Note that this may end up consuming a
+			// block notification for the client, but given that we
+			// can no longer satisfy their other requests, this
+			// should be fine.
+			select {
+			case _, ok := <-client:
+				if !ok {
+					continue
+				}
+			default:
+			}
+			close(client)
+		}
+	}
+}
+
+// Request submits a request to retrieve the information of the given blocks.
+func (s *NetworkBlockSource) Request(
+	blocks ...*chainhash.Hash) (<-chan *wire.MsgBlock, error) {
+
+	// Queue the channel with as many blocks as we expect.
+	blockChan := make(chan *wire.MsgBlock, len(blocks))
+
+	select {
+	case s.blockReqs <- &blockReq{
+		blocks:    blocks,
+		blockChan: blockChan,
+	}:
+	case <-s.quit:
+		return nil, errors.New("shutting down")
+	}
+
+	return blockChan, nil
+}
+
+// requestHandler represents the NetworkBlockSource's main event loop where
+// incoming requests are handled, queried, and notified.
+func (s *NetworkBlockSource) requestHandler() {
+	defer s.wg.Done()
+
+	blockChan := make(chan *wire.MsgBlock)
+	var peers []btcjson.GetPeerInfoResult
+	defer s.cfg.CyclePeerTicker.Stop()
+out:
+	for {
+		// We'll wait to receive at least one block request before doing
+		// any work.
+		if len(s.activeReqs) == 0 {
+			select {
+			case req := <-s.blockReqs:
+				log.Debugf("Received new request for block(s) "+
+					"%v", req.blocks)
+				for _, block := range req.blocks {
+					s.activeReqs[*block] = append(
+						s.activeReqs[*block], req.blockChan,
+					)
+				}
+
+			case <-s.quit:
+				return
+			}
+		}
+
+		// If this is our first iteration or we've exhausted all of our
+		// current peers, refresh them.
+		if len(peers) == 0 {
+			log.Debugf("Refreshing list of candidate bitcoind peers")
+
+			var err error
+			peers, err = s.refreshCandidatePeers()
+			switch {
+			case err != nil:
+				log.Errorf("Failed to refresh set of candidate "+
+					"bitcoind peers: %v", err)
+				fallthrough
+
+			case len(peers) == 0:
+				log.Debugf("No candidate bitcoind peers "+
+					"found, waiting %v to retry",
+					s.cfg.RefreshPeersBackoff)
+
+				select {
+				// Try again after our interval elapses.
+				case <-time.After(s.cfg.RefreshPeersBackoff):
+					continue
+				case <-s.quit:
+					return
+				}
+			}
+
+			log.Debugf("Found %v candidate bitcoind peers", len(peers))
+		}
+
+		// With our set of candidate peers refreshed, we'll start with
```

Comment:
So I take it attempting to re-use the existing query interface added to neutrino didn't end up being workable? It already supports being able to multi-plex several requests across a fixed number of peers. cc @halseth

Reply:
For reference: https://github.com/lightninglabs/neutrino/tree/master/query

Reply:
It even has a peer rank abstraction that can be used to base prioritization on latency, then punish/rewards those tht actually respond quickly.